modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in

```diff
@@ -223,56 +223,105 @@ struct fast_length_untyped_helper<T, true> {
 };
 
 template <typename T>
-typename TypeTraits<T>::ElementType fast_length(const T& p) {
+typename TypeTraits<T>::ElementType fast_length(const T &p) {
   return fast_length_untyped_helper<T>::_(p);
 }
 
 template <typename T>
-typename TypeTraits<T>::ElementType fast_distance(const T& x, const T& y) {
+typename TypeTraits<T>::ElementType fast_distance(const T &x, const T &y) {
   return geometric::fast_length<T>(x - y);
 }
 
 template <typename T, typename E = typename TypeTraits<T>::ElementType>
 struct fast_normalize_helper {
-  static T _(const T& p) { return p * __abacus_rsqrt(geometric::dot<T>(p, p)); }
+  static T _(const T &p) { return p * __abacus_rsqrt(geometric::dot<T>(p, p)); }
 };
 
 template <typename T>
 struct fast_normalize_helper<T, abacus_float> {
-  static T _(const T& p) {
+  static T _(const T &p) {
     return p * __abacus_half_rsqrt(geometric::dot<T>(p, p));
   }
 };
 
 template <typename T>
-T fast_normalize(const T& p) {
+T fast_normalize(const T &p) {
   return fast_normalize_helper<T>::_(p);
 }
 
 template <typename T>
-T normalize(const T& p) {
-  using IntType = typename TypeTraits<T>::SignedType;
-
-  // Check if *any* vector lane is INF.
-  const IntType isInf = __abacus_isinf(p);
-
-  // Broadcasts scalar abacus_int to vector type since `__abacus_any()` returns
-  // scalar 1 if any vector lane has a sign bit set.
-  const IntType isAnyInf = __abacus_any(isInf) == 1;
+T normalize(T p) {
+  constexpr size_t num_elements = TypeTraits<T>::num_elements;
+  static_assert(num_elements >= 1 && num_elements <= 16);
+  if constexpr (num_elements == 1) {  // NOLINT
+    return p > 0 ? 1 : p < 0 ? -1 : p;
+  } else {  // NOLINT
+    using ElementType = typename TypeTraits<T>::ElementType;
+    using UnsignedType = typename TypeTraits<T>::UnsignedType;
+    using ScalarUnsignedType = typename FPShape<T>::ScalarUnsignedType;
+
+    // Determine the largest element by absolute magnitude. Sign is ignored, and
+    // NaN compares as higher than any other value.
+    const UnsignedType au = detail::cast::as<UnsignedType>(p);
+    const UnsignedType abs_au = au & FPShape<T>::InverseSignMask();
+    ScalarUnsignedType max_mag_au = abs_au[0];
+    for (size_t i = 1; i < num_elements; ++i) {
+      ScalarUnsignedType absElem_au = abs_au[i];
+      if (max_mag_au < absElem_au) max_mag_au = absElem_au;
+    }
+    const ElementType max_mag = detail::cast::as<ElementType>(max_mag_au);
 
-  // Flush non-INFs to zero if any other lane is INF.
-  T infsFlushed = __abacus_select(p, T(0.0), __abacus_isfinite(p) & isAnyInf);
+    if (max_mag == 0) {
+      // If all elements were zero, handle this specially by returning the
+      // vector unmodified.
+      return p;
+    }
 
-  // Flush positive INF to 1.
-  const IntType isPosInf = isInf & ~__abacus_signbit(p);
-  infsFlushed = __abacus_select(infsFlushed, T(1.0), isPosInf);
+    if (max_mag == 0) {
```

Comment:
is this not the same as L274?

Reply:
It is, thanks, I think I messed up the conflict resolution after a rebase, will remove.